torch/utils/data/dataloader.py

```diff
@@ -122,30 +147,38 @@ def _worker_loop(dataset, index_queue, data_queue, done_event, collate_fn, seed,
         pass
 
 
-def _pin_memory_loop(in_queue, out_queue, done_event, pin_memory, device_id):
-    if pin_memory:
-        torch.cuda.set_device(device_id)
+def _pin_memory_loop(in_queue, out_queue, device_id, done_event):
+    torch.cuda.set_device(device_id)
 
+    # See NOTE [ Data Loader Multiprocessing Shutdown Logic ] for details on the
+    # logic of this function.
     while True:
         try:
-            r = in_queue.get()
+            r = in_queue.get(timeout=MP_STATUS_CHECK_INTERVAL)
+        except queue.Empty:
+            continue
         except Exception:
             if done_event.is_set():
-                return
+                # Weird things can happen when shutting down, e.g., fd being
+                # closed when tensors are shared via fds.
```

Comment:
How does this apply to this branch specifically?

Reply:
It is reading from `worker_result_queue` so in exiting it might throw `ConnectionResetByPeer`, and `IOError: file is closed` IIRC.